### DISCUSSION TITLE
Conditional params parsing

### DISCUSSION THREAD

**Discussion by Scrubles** [Mar-7th/StarRailRes#13] [General]
Hey @mobyw, I opened another thread as you instructed.

About your example, can you give more details?

I was looking at this example (DHIL signature LC).

`   
 "1": {
      "SkillID": 23015,
      "SkillName": {
        "Hash": 703818376
      },
      "SkillDesc": {
        "Hash": 1374581526
      },
      "Level": 1,
      "AbilityName": "Ability23015",
      "ParamList": [
        {
          "Value": 0.18
        },
        {
          "Value": 2
        },
        {
          "Value": 2
        },
        {
          "Value": 0.18
        },
        {
          "Value": 0.06
        }
      ],
      "AbilityProperty": [
        {
          "PropertyType": "CriticalChanceBase",
          "Value": {
            "Value": 0.18
          }
        }
      ]
    }
`

It has 3 dynamic values, the max stack, the energy regen and the additional ATK%

On the EquipmemtAbility.json I found these:

`
"MEquip_23015_Sub": {
  "MaxLayer": 3,
  "LayerAddWhenStack": {
    "IsDynamic": false,
    "FixedValue": {
      "Value": 1
    }
  },
  "_CallbackList": [
    {
      "Event": "OnStack",
      "CallbackConfig": [
        {
          "$type": "RPG.GameCore.SetDynamicValueByModifierValue",
          "ReadTargetType": {
            "$type": "RPG.GameCore.TargetAlias",
            "Alias": "ModifierOwnerEntity"
          },
          "ValueType": "Layer",
          "Multiplier": {
            "IsDynamic": false,
            "FixedValue": {
              "Value": 1
            }
          },
          "DynamicKey": "_Layer"
        },
        {
          "$type": "RPG.GameCore.StackProperty",
          "TargetType": {
            "$type": "RPG.GameCore.TargetAlias",
            "Alias": "ModifierOwnerEntity"
          },
          "Property": "AttackAddedRatio",
          "PropertyValue": {
            "IsDynamic": true,
            "PostfixExpr": {
              "OpCodes": "AQABAQQK",
              "FixedValues": [],
              "DynamicHashes": [
                -1315444205,
                1912601768
              ]
            }
          }
        },
        {
          "$type": "RPG.GameCore.StackProperty",
          "TargetType": {
            "$type": "RPG.GameCore.TargetAlias",
            "Alias": "ModifierOwnerEntity"
          },
          "Property": "SPRatioBase",
          "PropertyValue": {
            "IsDynamic": true,
            "PostfixExpr": {
              "OpCodes": "AQABAQQK",
              "FixedValues": [],
              "DynamicHashes": [
                265093437,
                1912601768
              ]
            }
          }
        }
      ]
    }
  ],
  "Stacking": "ReplaceByCaster",
  "DynamicValues": {
    "Values": {
      "1912601768": {}
    }
  }
}
`

But their ReadInfo is like this:

`
"-1970381737": {
  "ReadInfo": {
    "Type": "SkillTreeParam",
    "Str": ""
  }
},
"615861046": {
  "ReadInfo": {
    "Type": "SkillRank",
    "Str": ""
  }
},
"-1315444205": {
  "ReadInfo": {
    "Type": "BattleEvent",
    "Str": ""
  }
},
"265093437": {
  "ReadInfo": {
    "Type": 8,
    "Str": ""
  }
}
`
Maybe this is a bad example and I didn't find their connection, but then I found this on StatusConfig.json

`
"40230151": {
  "StatusID": 40230151,
  "ModifierName": "MEquip_23015_Sub",
  "StatusName": {
    "Hash": 1020414610
  },
  "StatusType": "Buff",
  "StatusDesc": {
    "Hash": 112084872
  },
  "StatusIconPath": "SpriteOutput/BuffIcon/Inlevel/IconBuffAttackUp.png",
  "StatusIconPathHighSize": "",
  "StatusEffect": {
    "Hash": 371857150
  },
  "CanDispel": true,
  "ReadParamList": [
    "#SkillEquip_P4_AttackAddedRatio",
    "#SkillEquip_P5_SPRatio",
    "#SkillEquip_P3_MaxLayer"
  ]
}
`

I tought I was goind somewhere, but then I checked an harder example, DHIL E1.
In AvatarStatusConfig I found this:

`
"10012133": {
  "StatusID": 10012133,
  "ModifierName": "MAvatar_DanHengIL_00_AllDamageTypeAddedRatio",
  "StatusName": {
    "Hash": -1409722473
  },
  "StatusType": "Buff",
  "StatusDesc": {
    "Hash": -881384327
  },
  "StatusIconPath": "SpriteOutput/BuffIcon/Inlevel/IconBuffAttackUpElement.png",
  "StatusIconPathHighSize": "",
  "StatusEffect": {
    "Hash": 371857150
  },
  "CanDispel": true,
  "ReadParamList": [
    "MDF_PropertyValue",
    "MDF_MaxLayer"
  ]
}
`

When I went to check MAvatar_DanHengIL_00_AllDamageTypeAddedRatio, I got stuck. Maybe their ReadInfo is elsewhere.

On a completely different topic, do you know if there's data from the leak?

Also, about the generating scripts, that would be much appreciated.

**Comment by mobyw** [comment]
Hi @Scrubles. I have converted the origin issue into this discussion.

Leaked data, or beta data will not be updated to this repo. They may have changes before they are officially released. If you are interested in these, you can find the corresponding community to obtain them. Data of his repo will be updated after StarRailData is updated, usually some time after the official version is available for pre-download.

In order to make it clear for others to read, only the issue of params parsing will be discussed in this thread. For other questions, just open a new issue or discussion.

**Comment by mobyw** [comment]
For LC, skill info can be found at`ExcelOutput/EquipmentSkillConfig.json`, here is the Superimposition 1 part of "Brighter Than the Sun":

```json
{
  "SkillID": 23015,
  "SkillName": { "Hash": 703818376 },
  "SkillDesc": { "Hash": 1374581526 },
  "Level": 1,
  "AbilityName": "Ability23015",
  "ParamList": [
    { "Value": 0.18 },
    { "Value": 2 },
    { "Value": 2 },
    { "Value": 0.18 },
    { "Value": 0.06 }
  ],
  "AbilityProperty": [
    {
      "PropertyType": "CriticalChanceBase",
      "Value": { "Value": 0.18 }
    }
  ]
}
```

There are 5 params in `ParamList`, in the subsequent config, it will be obtained according to the index (0, 1, 2, 3, 4).

`Ability23015` is its ability id, which is defined in `Config/ConfigAbility/EquipmemtAbility.json`, `Ability23015` contents:

```jsonc
{
  "Modifiers": {
    // hide
  },
  "Name": "Ability23015",
  "TargetInfo": {
    "TargetType": "Caster"
  },
  "OnStart": [
    // hide
  ],
  "DynamicValues": {
    "Values": {
      "-1970381737": { "ReadInfo": { "Type": "SkillTreeParam", "Str": "" } },
      "615861046": { "ReadInfo": { "Type": "SkillRank", "Str": "" } },
      "-1315444205": { "ReadInfo": { "Type": "BattleEvent", "Str": "" } },
      "265093437": { "ReadInfo": { "Type": 8, "Str": "" } }
    }
  }
}
```

There are 4 values in `DynamicValues`, obviously these values are related to LC params, and all have `ReadInfo`. (If there is no `ReadInfo`, it means that it is just a variable and not related to param.)

`Str` is the param source. `Str` of `ReadInfo` is empty for LC, because it must be from LC skill, this is different for character (can be from eidolons, skills, or traces).

`Type` is the param index, follow the rules below:

```python
{
  "None": 0,
  "SkillTreeParam": 1,
  "SkillRank": 2,
  "BattleEvent": 3,
  8: 4,
  10: 5,
  12: 6,
}
```

The above 4 values correspond to params index respectively: 1, 2, 3, 4.

Index 0 does not require additional conditions, so it will not appear here.

**Reply by Scrubles** [reply]
That rule seems very random, thanks for pointing that out. I'll try to make a connection here.
